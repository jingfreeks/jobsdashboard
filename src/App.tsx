import { Routes, Route, Navigate } from 'react-router-dom'
import { Login, Register, Dashboard } from './pages'
import ProtectedRoute from './components/ProtectedRoute'
import './App.css'

function App() {
  return (
    <Routes>
      <Route path="/login" element={<Login />} />
      <Route path="/register" element={<Register />} />
      <Route 
        path="/dashboard" 
        element={
          <ProtectedRoute>
            <Dashboard />
          </ProtectedRoute>
        } 
      />
      <Route path="*" element={<Navigate to="/login" replace />} />
    </Routes>
  )
}

export default App
