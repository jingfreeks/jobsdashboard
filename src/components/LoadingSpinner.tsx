import React from 'react';

const LoadingSpinner: React.FC = () => {
  return (
    <div className="min-h-screen flex items-center justify-center bg-slate-100">
      <div className="flex flex-col items-center gap-4">
        <div className="animate-spin rounded-full h-12 w-12 border-b-2 border-blue-600"></div>
        <p className="text-slate-600 font-medium">Loading...</p>
      </div>
    </div>
  );
};

export default LoadingSpinner; 