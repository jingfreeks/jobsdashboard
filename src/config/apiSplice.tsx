/* eslint-disable @typescript-eslint/no-explicit-any */
import {setCredentials, setLogout} from '@/features/auth';
import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';
const baseUrl = import.meta.env.VITE_API_BASE_URL;
const baseQuery = fetchBaseQuery({
  baseUrl,
  credentials: 'include',
   
  prepareHeaders: (headers, {getState}: any) => {
    const token = getState().auth.token;
    if (token) {
      headers.set('authorization', `Bearer ${token}`);
    }
    return headers;
  },
});

const baseQueryWithAuth = async (
  args: string,
  api: any,
  extraOptions: string,
) => {
  let result: any = await baseQuery(args, api, extraOptions);
  console.log(result);
  if (result?.error?.originalStatus === 403) {
    const refereshResult = await baseQuery('/refresh', api, extraOptions);

    if (refereshResult?.data) {
      const currentAuth = api.getState().auth;
      //store the new token
      api.dispatch(setCredentials({
        user: currentAuth.user,
        accessToken: (refereshResult.data as any).accessToken,
        userId: currentAuth.userId,
        roles: currentAuth.roles
      }));
      //retry the original query with new access token
      result = await baseQuery(args, api, extraOptions);
    } else {
      api.dispatch(setLogout());
    }
  }
  return result;
};

export const apiSlice = createApi<any, any>({
  baseQuery: baseQueryWithAuth,
  reducerPath: 'api',
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  endpoints: _ => ({}),
  tagTypes: ['City', 'Jobs', 'MyJobs', 'JobDetails', 'Profile'] as any,
});