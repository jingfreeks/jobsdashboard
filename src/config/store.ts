import { configureStore } from '@reduxjs/toolkit';
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import authReducer from '@/features/auth';
import {apiSlice} from './apiSplice';

// Configure persistence for auth reducer
const authPersistConfig = {
  key: 'auth',
  storage,
  whitelist: ['user', 'token', 'userId', 'roles'], // Only persist these fields
  timeout: 10000, // 10 second timeout for storage operations
  serialize: true, // Enable serialization
  deserialize: true, // Enable deserialization
};

const persistedAuthReducer = persistReducer(authPersistConfig, authReducer);

export const store = configureStore({
  reducer: {
    auth: persistedAuthReducer,
    [apiSlice.reducerPath]: apiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
        ignoredPaths: ['auth'], // Ignore auth state for serialization checks
      },
      immutableCheck: {
        ignoredPaths: ['auth'], // Ignore auth state for immutability checks
      },
    }).concat(apiSlice.middleware),
  devTools: process.env.NODE_ENV !== 'production',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch; 