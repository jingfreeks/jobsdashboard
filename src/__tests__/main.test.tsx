import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { render } from '@testing-library/react';
import { createRoot } from 'react-dom/client';
import { StrictMode } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
import { store, persistor } from '../config/store';
import LoadingSpinner from '../components/LoadingSpinner';
import App from '../App';

// Mock React DOM
vi.mock('react-dom/client', () => ({
  createRoot: vi.fn(() => ({
    render: vi.fn(),
  })),
}));

// Mock the store and persistor
vi.mock('../config/store', () => ({
  store: {
    getState: vi.fn(() => ({})),
    dispatch: vi.fn(),
    subscribe: vi.fn(),
  },
  persistor: {
    purge: vi.fn(),
    flush: vi.fn(),
    pause: vi.fn(),
    persist: vi.fn(),
    resume: vi.fn(),
    subscribe: vi.fn(() => vi.fn()), // Return unsubscribe function
    getState: vi.fn(() => ({ bootstrapped: true })),
  },
}));

// Mock the App component
vi.mock('../App', () => ({
  default: vi.fn(() => <div data-testid="app-component">App Component</div>),
}));

// Mock the LoadingSpinner component
vi.mock('../components/LoadingSpinner', () => ({
  default: vi.fn(() => <div data-testid="loading-spinner">Loading...</div>),
}));

// Mock CSS imports
vi.mock('../index.css', () => ({}));

// Mock document.getElementById
const mockRootElement = document.createElement('div');
mockRootElement.id = 'root';

describe('Main Entry Point', () => {
  let mockCreateRoot: any;
  let mockRender: any;

  beforeEach(() => {
    vi.clearAllMocks();
    
    // Setup DOM
    document.body.appendChild(mockRootElement);
    
    // Setup mocks
    mockRender = vi.fn();
    mockCreateRoot = vi.mocked(createRoot);
    mockCreateRoot.mockReturnValue({
      render: mockRender,
    });
  });

  afterEach(() => {
    // Cleanup
    if (document.body.contains(mockRootElement)) {
      document.body.removeChild(mockRootElement);
    }
  });

  describe('DOM Element Setup', () => {
    it('finds root element correctly', () => {
      expect(document.getElementById('root')).toBe(mockRootElement);
    });

    it('root element exists in document', () => {
      expect(mockRootElement).toBeInTheDocument();
      expect(mockRootElement.id).toBe('root');
    });

    it('root element is a div', () => {
      expect(mockRootElement.tagName).toBe('DIV');
    });
  });

  describe('React 18 createRoot', () => {
    it('calls createRoot with root element', () => {
      // Test that createRoot is called with the root element
      mockCreateRoot(mockRootElement);
      
      expect(mockCreateRoot).toHaveBeenCalledWith(mockRootElement);
      expect(mockCreateRoot).toHaveBeenCalledTimes(1);
    });

    it('createRoot returns object with render method', () => {
      const root = mockCreateRoot(mockRootElement);
      expect(root).toHaveProperty('render');
      expect(typeof root.render).toBe('function');
    });
  });

  describe('Component Structure', () => {
    it('renders with StrictMode wrapper', () => {
      const TestComponent = () => (
        <StrictMode>
          <div data-testid="strict-mode-content">Content</div>
        </StrictMode>
      );
      
      const { getByTestId } = render(<TestComponent />);
      expect(getByTestId('strict-mode-content')).toBeInTheDocument();
    });

    it('renders with Redux Provider', () => {
      const TestComponent = () => (
        <Provider store={store}>
          <div data-testid="redux-provider-content">Content</div>
        </Provider>
      );
      
      const { getByTestId } = render(<TestComponent />);
      expect(getByTestId('redux-provider-content')).toBeInTheDocument();
    });

    it('renders with PersistGate', () => {
      const TestComponent = () => (
        <PersistGate loading={<LoadingSpinner />} persistor={persistor}>
          <div data-testid="persist-gate-content">Content</div>
        </PersistGate>
      );
      
      const { getByTestId } = render(<TestComponent />);
      expect(getByTestId('persist-gate-content')).toBeInTheDocument();
    });

    it('renders with BrowserRouter', () => {
      const TestComponent = () => (
        <BrowserRouter>
          <div data-testid="browser-router-content">Content</div>
        </BrowserRouter>
      );
      
      const { getByTestId } = render(<TestComponent />);
      expect(getByTestId('browser-router-content')).toBeInTheDocument();
    });

    it('renders App component', () => {
      const { getByTestId } = render(<App />);
      expect(getByTestId('app-component')).toBeInTheDocument();
    });
  });

  describe('Component Integration', () => {
    it('integrates all components correctly', () => {
      const TestComponent = () => (
        <StrictMode>
          <Provider store={store}>
            <PersistGate loading={<LoadingSpinner />} persistor={persistor}>
              <BrowserRouter>
                <App />
              </BrowserRouter>
            </PersistGate>
          </Provider>
        </StrictMode>
      );
      
      const { getByTestId } = render(<TestComponent />);
      
      // All components should be rendered
      expect(getByTestId('app-component')).toBeInTheDocument();
    });

    it('LoadingSpinner is used as PersistGate loading prop', () => {
      const TestComponent = () => (
        <PersistGate loading={<LoadingSpinner />} persistor={persistor}>
          <div>Content</div>
        </PersistGate>
      );
      
      const { container } = render(<TestComponent />);
      // Test that the component renders without errors
      expect(container.firstChild).toBeInTheDocument();
    });

    it('App component is rendered inside BrowserRouter', () => {
      const TestComponent = () => (
        <BrowserRouter>
          <App />
        </BrowserRouter>
      );
      
      const { getByTestId } = render(<TestComponent />);
      expect(getByTestId('app-component')).toBeInTheDocument();
    });
  });

  describe('Redux Store Integration', () => {
    it('store is properly configured', () => {
      expect(store).toBeDefined();
      expect(store.getState).toBeDefined();
      expect(store.dispatch).toBeDefined();
      expect(store.subscribe).toBeDefined();
    });

    it('persistor is properly configured', () => {
      expect(persistor).toBeDefined();
      expect(persistor.purge).toBeDefined();
      expect(persistor.flush).toBeDefined();
      expect(persistor.pause).toBeDefined();
      expect(persistor.persist).toBeDefined();
      //expect(persistor.resume).toBeDefined();
    });

    it('Provider receives store prop', () => {
      const TestComponent = () => (
        <Provider store={store}>
          <div>Content</div>
        </Provider>
      );
      
      const { container } = render(<TestComponent />);
      expect(container.firstChild).toBeInTheDocument();
    });

    it('PersistGate receives persistor prop', () => {
      const TestComponent = () => (
        <PersistGate loading={<LoadingSpinner />} persistor={persistor}>
          <div>Content</div>
        </PersistGate>
      );
      
      const { container } = render(<TestComponent />);
      expect(container.firstChild).toBeInTheDocument();
    });
  });

  describe('CSS and Styling', () => {
    it('imports index.css correctly', () => {
      // The CSS import should not cause errors
      expect(() => {
        require('../index.css');
      }).not.toThrow();
    });

    it('CSS imports are handled properly', () => {
      // Mock CSS import should work
      const cssModule = require('../index.css');
      expect(cssModule).toBeDefined();
    });
  });

  describe('Error Handling', () => {
    it('handles missing root element gracefully', () => {
      // Remove root element
      if (document.body.contains(mockRootElement)) {
        document.body.removeChild(mockRootElement);
      }
      
      // Should not throw when root element is missing
      expect(() => {
        document.getElementById('root');
      }).not.toThrow();
    });

    it('handles createRoot errors gracefully', () => {
      mockCreateRoot.mockImplementation(() => {
        throw new Error('createRoot error');
      });
      
      expect(() => {
        mockCreateRoot(mockRootElement);
      }).toThrow('createRoot error');
    });

    it('handles render errors gracefully', () => {
      mockRender.mockImplementation(() => {
        throw new Error('render error');
      });
      
      expect(() => {
        mockRender(<div>Test</div>);
      }).toThrow('render error');
    });
  });

  describe('Performance and Memory', () => {
    it('does not cause memory leaks with frequent renders', () => {
      const root = mockCreateRoot(mockRootElement);
      
      // Simulate frequent renders
      for (let i = 0; i < 100; i++) {
        root.render(<div>Test {i}</div>);
      }
      
      expect(mockRender).toHaveBeenCalledTimes(100);
    });

    it('handles large component trees efficiently', () => {
      const LargeComponent = () => (
        <div>
          {Array.from({ length: 1000 }, (_, i) => (
            <div key={i}>Item {i}</div>
          ))}
        </div>
      );
      
      const root = mockCreateRoot(mockRootElement);
      root.render(<LargeComponent />);
      
      expect(mockRender).toHaveBeenCalledTimes(1);
    });
  });

  describe('Accessibility', () => {
    it('maintains proper document structure', () => {
      expect(document.documentElement).toBeInTheDocument();
      expect(document.body).toBeInTheDocument();
      expect(document.head).toBeInTheDocument();
    });

    it('root element has proper attributes', () => {
      expect(mockRootElement.id).toBe('root');
      expect(mockRootElement.tagName).toBe('DIV');
    });

    it('renders without accessibility violations', () => {
      const TestComponent = () => (
        <div role="main">
          <h1>Main Content</h1>
        </div>
      );
      
      const { getByRole } = render(<TestComponent />);
      expect(getByRole('main')).toBeInTheDocument();
    });
  });

  describe('React 18 Features', () => {
    it('uses createRoot instead of ReactDOM.render', () => {
      expect(mockCreateRoot).toBeDefined();
      expect(typeof mockCreateRoot).toBe('function');
    });

    it('createRoot is called with correct element', () => {
      mockCreateRoot(mockRootElement);
      expect(mockCreateRoot).toHaveBeenCalledWith(mockRootElement);
    });

    it('render method is called on root object', () => {
      const root = mockCreateRoot(mockRootElement);
      const testElement = <div>Test</div>;
      root.render(testElement);
      
      expect(mockRender).toHaveBeenCalledWith(testElement);
    });
  });

  describe('Component Props and Configuration', () => {
    it('StrictMode wraps entire application', () => {
      const TestComponent = () => (
        <StrictMode>
          <div data-testid="strict-mode-wrapper">Content</div>
        </StrictMode>
      );
      
      const { getByTestId } = render(<TestComponent />);
      expect(getByTestId('strict-mode-wrapper')).toBeInTheDocument();
    });

    it('Provider wraps with store', () => {
      const TestComponent = () => (
        <Provider store={store}>
          <div data-testid="provider-wrapper">Content</div>
        </Provider>
      );
      
      const { getByTestId } = render(<TestComponent />);
      expect(getByTestId('provider-wrapper')).toBeInTheDocument();
    });

    it('PersistGate has correct props', () => {
      const TestComponent = () => (
        <PersistGate loading={<LoadingSpinner />} persistor={persistor}>
          <div data-testid="persist-gate-wrapper">Content</div>
        </PersistGate>
      );
      
      const { getByTestId } = render(<TestComponent />);
      expect(getByTestId('persist-gate-wrapper')).toBeInTheDocument();
    });

    it('BrowserRouter provides routing context', () => {
      const TestComponent = () => (
        <BrowserRouter>
          <div data-testid="router-wrapper">Content</div>
        </BrowserRouter>
      );
      
      const { getByTestId } = render(<TestComponent />);
      expect(getByTestId('router-wrapper')).toBeInTheDocument();
    });
  });

  describe('Snapshot Testing', () => {
    it('matches main component structure snapshot', () => {
      const TestComponent = () => (
        <StrictMode>
          <Provider store={store}>
            <PersistGate loading={<LoadingSpinner />} persistor={persistor}>
              <BrowserRouter>
                <App />
              </BrowserRouter>
            </PersistGate>
          </Provider>
        </StrictMode>
      );
      
      const { container } = render(<TestComponent />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches individual component snapshots', () => {
      const { container: appContainer } = render(<App />);
      expect(appContainer.firstChild).toMatchSnapshot();
      
      const { container: loadingContainer } = render(<LoadingSpinner />);
      expect(loadingContainer.firstChild).toMatchSnapshot();
    });
  });

  describe('Integration Testing', () => {
    it('full application renders without errors', () => {
      const FullApp = () => (
        <StrictMode>
          <Provider store={store}>
            <PersistGate loading={<LoadingSpinner />} persistor={persistor}>
              <BrowserRouter>
                <App />
              </BrowserRouter>
            </PersistGate>
          </Provider>
        </StrictMode>
      );
      
      const { getByTestId } = render(<FullApp />);
      expect(getByTestId('app-component')).toBeInTheDocument();
    });

    it('components work together correctly', () => {
      const TestApp = () => (
        <Provider store={store}>
          <BrowserRouter>
            <App />
          </BrowserRouter>
        </Provider>
      );
      
      const { getByTestId } = render(<TestApp />);
      expect(getByTestId('app-component')).toBeInTheDocument();
    });
  });

  describe('Environment and Configuration', () => {
    it('works in test environment', () => {
      expect(process.env.NODE_ENV).toBeDefined();
    });

    it('handles different environments correctly', () => {
      const originalEnv = process.env.NODE_ENV;
      
      // Test development environment
      process.env.NODE_ENV = 'development';
      expect(process.env.NODE_ENV).toBe('development');
      
      // Test production environment
      process.env.NODE_ENV = 'production';
      expect(process.env.NODE_ENV).toBe('production');
      
      // Restore original
      process.env.NODE_ENV = originalEnv;
    });

    it('imports work correctly', () => {
      expect(StrictMode).toBeDefined();
      expect(createRoot).toBeDefined();
      expect(BrowserRouter).toBeDefined();
      expect(Provider).toBeDefined();
      expect(PersistGate).toBeDefined();
      expect(store).toBeDefined();
      expect(persistor).toBeDefined();
      expect(LoadingSpinner).toBeDefined();
      expect(App).toBeDefined();
    });
  });
}); 