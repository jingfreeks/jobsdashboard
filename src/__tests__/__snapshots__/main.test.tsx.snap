// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Main Entry Point > Snapshot Testing > matches individual component snapshots 1`] = `
<div
  data-testid="app-component"
>
  App Component
</div>
`;

exports[`Main Entry Point > Snapshot Testing > matches individual component snapshots 2`] = `
<div
  data-testid="loading-spinner"
>
  Loading...
</div>
`;

exports[`Main Entry Point > Snapshot Testing > matches main component structure snapshot 1`] = `
<div
  data-testid="app-component"
>
  App Component
</div>
`;
